tools/yield/Common.cpp

```diff
@@ -0,0 +1,91 @@
+/// \file Common.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/Model/RawBinaryView.h"
+#include "revng/Pipes/Kinds.h"
+#include "revng/Support/MetaAddress/YAMLTraits.h"
+#include "revng/Support/YAMLTraits.h"
+
+#include "Common.h"
+
+namespace options {
+
+using namespace llvm::cl;
+
+OptionCategory ThisToolCategory("Tool options", "");
+
+static opt<std::string> BinaryPath(llvm::cl::Positional,
+                                   llvm::cl::Required,
+                                   desc("<input binary>"),
+                                   value_desc("binary"),
+                                   cat(ThisToolCategory));
+
+static opt<std::string> LLVMModulePath(llvm::cl::Positional,
+                                       llvm::cl::Required,
+                                       desc("<input llvm-module>"),
+                                       value_desc("llvm-module"),
+                                       cat(ThisToolCategory));
+
+static opt<std::string> ModelPath(llvm::cl::Positional,
+                                  desc("<input model>"),
+                                  value_desc("model"),
+                                  init("-"),
+                                  cat(ThisToolCategory));
+
+} // namespace options
+
+static const model::Binary &
+findTheModel(const ModelInModule &Module,
+             llvm::Optional<ModelInModule> &MaybeModel) {
+  if (!MaybeModel) {
+    // Explicit model path is not usable, use the model from the ir.
+    return *Module.getModel();
+  } else {
+    // A valid path is provided - use the model from there.
+    return *MaybeModel->getModel();
+  }
+}
+
+static revng::pipes::StringMapContainer createMap() {
+  using C = revng::pipes::StringMapContainer;
+  return C("", "", pipeline::Kind{ "", &revng::pipes::RootRank });
+}
+
+ReturnValueType parseCommandLineOptions(int Argc, char *Argv[]) {
+  llvm::cl::HideUnrelatedOptions({ &options::ThisToolCategory });
+  llvm::cl::ParseCommandLineOptions(Argc, Argv);
+
+  llvm::ExitOnError ExitOnError;
+  ObjectLifetimeController Result;
+
+  // Load the llvm module
+  Result.IR = ExitOnError(ModelInModule::loadModule(options::LLVMModulePath));
+  revng_assert(Result.IR.hasModule());
+
+  // Locate and load the model
+  using namespace llvm;
+  Result.Model = expectedToOptional(ModelInModule::load(options::ModelPath));
+  const model::Binary &Model = findTheModel(Result.IR, Result.Model);
```

Comment:
`ModelInModule` is useful when you don't really care about the LLVM module, in this case it's probably better to 1) just manually parse the IR, 2) `if (hasModule(M)) Model = loadModel(M); else Model = deserialize(options::ModelPath)`.

Reply:
What about something like this then: https://github.com/revng/revng/pull/173/commits/6ba47650c5d095ef1c875456ca1e62a040e18277?

Reply:
I was suggesting not to use `ModelInModule` entirely, since it's designed to serve a different use case.

But this is not too bad.

Reply:
When I see "just manually parse X", I'd rather go any other way I can possibly go with :D

At the very least this simplifies the support this tool ~~I never even wanted to create~~ will possibly need since it relies on a well-supported part of the project instead of hand-rolling a parser for a changing format.

Reply:
I have no idea why, but something like `ModelInModule` gives more of a "stable-API-that-is-definitely-not-going-anywhere" feeling to it. I guess I shouldn't make decisions like this based on hunches, but yeah, either way works!

Reply:
Either we extend `ModelInModule` to handle the situation without having to call `ModelInModule::load*` twice, or let's do things by hand as I suggested.

Reply:
Gotcha.

Reply:
Something like this?

https://github.com/revng/revng/pull/173/commits/5a4b366e115afac428ba2de58dd13194d1e7ad6b